### DISCUSSION TITLE
Node-API: RegExp support

### DISCUSSION THREAD

**Discussion by metafloor** [nodejs/help#4768] [Ideas]
In the process of writing some data marshalling code, I noticed that napi does not have support for RegExp.  Is this intentional due to a technical issue?  Or just something that hasn't been done yet?

I am willing to write the code (and documentation) for the following proposed RegExp support.  The naming conventions are modeled after the similar string functions.

```c
napi_status napi_create_regexp_utf8(
    napi_env env, 
    const char* source, 
    size_t source_length, 
    const char* flags, 
    size_t flags_length, 
    napi_value* res);

napi_status napi_create_regexp_utf16(
    napi_env env, 
    const char16_t* source, 
    size_t source_length, 
    const char* flags,    // should flags be char16_t?
    size_t flags_length, 
    napi_value* res);

napi_status napi_is_regexp(
    napi_env env,
    napi_value value,
    bool* res);

napi_status napi_get_value_regexp_utf8(
    napi_env env,
    napi_value value,
    char** source,
    size_t* source_length,  // [in,out] parameter
    char** flags,
    size_t* flags_length);  // [in,out] parameter

napi_status napi_get_value_regexp_utf16(
    napi_env env,
    napi_value value,
    char16_t** source,
    size_t* source_length,  // [in,out] parameter
    char** flags,
    size_t* flags_length);  // [in,out] parameter
```

The `napi_get_value_regexp_*` functions would support the usual convention of returning the length of the strings when `source` and `flags` are NULL.

The flags are represented as strings even though v8 uses a bitmask, and should provide better portability to other platforms.

Any thoughts on this would be appreciated.

**Comment by bnoordhuis** [comment]
One of n-api's design goals is making it easy to port over to other JS engines and that constrains the API surface. Taking quickjs as a, ah, quick litmus test, I don't think it exposes regexp functionality through its C API.
